### PR TITLE
[Flow EVM] adding state update checksum to transaction execution events

### PR DESCRIPTION
**Pull request by ramtinms**
This adds a new field (`stateUpdateChecksum`) to transaction execution events, that can be used to validate the state changes when a transaction is re-executed off-chain. 
the value of this checksum is the first 4 bytes of the commitment over the state update (4 bytes are used as checksum to save event space and minimize the cost on tx execution).

### CODE REVIEW

File: fvm/evm/events/utils.go [hunk 1]
```diff
@@ -30,3 +32,20 @@ func hashToCadenceArrayValue(hash gethCommon.Hash) cadence.Array {
...
+// checksumType is the Cadence type [UInt8;4]
+var checksumType = cadence.NewConstantSizedArrayType(ChecksumLength, cadence.UInt8Type)
+
+// checksumToCadenceArrayValue converts a checksum ([4]byte) into a Cadence array of type [UInt8;4]
+func checksumToCadenceArrayValue(checksum [4]byte) cadence.Array {
+	values := make([]cadence.Value, ChecksumLength)
+	for i := 0; i < ChecksumLength; i++ {
+		values[i] = cadence.NewUInt8(checksum[i])
+	}
+	fmt.Println(values)
```
Comment (turbolent) [inline fvm/evm/events/utils.go:48]:
```suggestion
```

**Review by turbolent — COMMENTED** [review]
Not familiar with the business logic changes (adding a new field), but Go code looks good

File: fvm/evm/events/utils.go (hunk 1 above)
Comment (ramtinms) [reply fvm/evm/events/utils.go]: fixed

File: fvm/evm/events/utils.go [hunk 2]
```diff
@@ -30,3 +30,19 @@ func hashToCadenceArrayValue(hash gethCommon.Hash) cadence.Array {
...
 }
+
+// ChecksumLength captures number of bytes a checksum uses
+const ChecksumLength = 4
+
+// checksumType is the Cadence type [UInt8;4]
+var checksumType = cadence.NewConstantSizedArrayType(ChecksumLength, cadence.UInt8Type)
+
+// checksumToCadenceArrayValue converts a checksum ([4]byte) into a Cadence array of type [UInt8;4]
+func checksumToCadenceArrayValue(checksum [4]byte) cadence.Array {
```
Comment (janezpodhostnik) [inline fvm/evm/events/utils.go:41]:
```suggestion
func checksumToCadenceArrayValue(checksum [ChecksumLength]byte) cadence.Array {
```

File: fvm/evm/events/events_test.go [hunk 3]
```diff
@@ -129,6 +131,7 @@ func TestEVMTransactionExecutedEventCCFEncodingDecoding(t *testing.T) {
 		assert.Equal(t, tep.GasConsumed, txResult.GasConsumed)
 		assert.Equal(t, tep.ErrorMessage, txResult.VMError.Error())
 		assert.Equal(t, tep.ReturnedData, txResult.ReturnedData)
+		assert.Equal(t, tep.StateUpdateChecksum[:], stateUpdateCommit[:4])
```
Comment (janezpodhostnik) [inline fvm/evm/events/events_test.go:134]:
```suggestion
		assert.Equal(t, tep.StateUpdateChecksum[:], stateUpdateCommit[:ChecksumLength])
```

File: fvm/evm/events/events.go [hunk 4]
```diff
@@ -188,6 +194,23 @@ func DecodeBlockEventPayload(event cadence.Event) (*BlockEventPayload, error) {
...
+	Payload             []byte          `cadence:"payload"`
+	ErrorCode           uint16          `cadence:"errorCode"`
+	GasConsumed         uint64          `cadence:"gasConsumed"`
+	ContractAddress     string          `cadence:"contractAddress"`
+	Logs                []byte          `cadence:"logs"`
+	BlockHeight         uint64          `cadence:"blockHeight"`
+	ErrorMessage        string          `cadence:"errorMessage"`
+	ReturnedData        []byte          `cadence:"returnedData"`
+	PrecompiledCalls    []byte          `cadence:"precompiledCalls"`
+	StateUpdateChecksum [4]byte         `cadence:"stateUpdateChecksum"`
```
Comment (janezpodhostnik) [inline fvm/evm/events/events.go:209]:
```suggestion
	StateUpdateChecksum [ChecksumLength]byte         `cadence:"stateUpdateChecksum"`
```

File: fvm/evm/events/events.go [hunk 5]
```diff
@@ -82,6 +82,11 @@ func (p *transactionEvent) ToCadence(chainID flow.ChainID) (cadence.Event, error
 
 	eventType := stdlib.CadenceTypesForChain(chainID).TransactionExecuted
 
+	// the first 4 bytes of StateChangeCommitment is used as checksum
+	var checksum [4]byte
```
Comment (janezpodhostnik) [inline fvm/evm/events/events.go:86]:
```suggestion
	var checksum [ChecksumLength]byte
```

**Review by janezpodhostnik — APPROVED** [review]

**Comment by janezpodhostnik** [comment]
Just out of curiosity, is there a reason the check sum length is 4?

**Comment by ramtinms** [comment]
@janezpodhostnik I tired to use a space similar to crc32 but it can actually be any size, thoughts on it?

**Review by sideninja — APPROVED** [review]

File: fvm/evm/events/utils.go (hunk 2 above)
Comment (ramtinms) [reply fvm/evm/events/utils.go:41]: updated

File: fvm/evm/events/events_test.go (hunk 3 above)
Comment (ramtinms) [reply fvm/evm/events/events_test.go:134]: updated

File: fvm/evm/events/events.go (hunk 5 above)
Comment (ramtinms) [reply fvm/evm/events/events.go:86]: updated

File: fvm/evm/events/events.go (hunk 4 above)
Comment (ramtinms) [reply fvm/evm/events/events.go:209]: updated

**Pull request merged**